gymnasium/envs/tabular/cliffwalking.py

```diff
@@ -0,0 +1,345 @@
+            jnp.maximum(jnp.minimum(new_position[0], 3),0),
+            jnp.maximum(jnp.minimum(new_position[1], 11),0)
+        ])
+    
+        #if we fell off, we have to start over from scratch from (3,0)
+        fallen = fell_off(new_position)
+        new_position = jnp.array([
+            new_position[0] * (1-fallen) + 3 * fallen, 
+            new_position[1] * (1-fallen)
+        ])
+        new_state = EnvState(
+            player_position = new_position.reshape((2,)),
+            last_action = action[0],
+            fallen = fallen
+        )
+
+        return new_state
+
+    def initial( self, rng: PRNGKey):
+        """Cliffwalking initial observation function."""
+
+        player_position = jnp.array([3,0])
+
+
+        state = EnvState(
+            player_position = player_position,
+            last_action = -1,
+            fallen = False
+        )
+
+        return state
+
+    def observation(self, state: EnvState) -> int:
+        """Cliffwalking observation."""
+        return jnp.array(state.player_position[0] * 12 + state.player_position[1]).reshape((1,))
+
+    def terminal(self, state: EnvState) -> jnp.ndarray:
+        """Determines if a particular Cliffwalking observation is terminal."""
+        return jnp.array_equal(state.player_position, jnp.array([3,11]))
+
+    def reward(self,
+         state: EnvState, action: ActType, next_state: StateType
+    ) -> jnp.ndarray:
+        """Calculates reward from a state."""
+        state = next_state
+        reward = -1  + (-99 * state.fallen[0])
+        return jax.lax.convert_element_type(reward, jnp.float32)
+
+    def render_init(self,
+         screen_width: int = 600, screen_height: int = 500
+    ) -> RenderStateType:
+        """Returns an initial render state."""
+        try:
+            import pygame
+        except ImportError:
+            raise DependencyNotInstalled(
+                "pygame is not installed, run `pip install gymnasium[classic_control]`"
+            )
+
+        rng = seeding.np_random(0)[0]
+
+        cell_size = (60, 60)
+        window_size = (
+            4 * cell_size[0],
+            12 * cell_size[1],
+        )
+
+        pygame.init()
+        screen = pygame.Surface((window_size[1], window_size[0]))
+
+        return screen
+
+    def render_image(
+        self,
+        state: StateType,
+        render_state: RenderStateType,
+    ) -> Tuple[RenderStateType, np.ndarray]:
+            """Renders an image from a state."""
+            try:
+                import pygame
+            except ImportError:
+                raise DependencyNotInstalled(
+                    "pygame is not installed, run `pip install gymnasium[toy_text]`"
+                )
+            window_surface = render_state
+
+
+
+            shape = (4,12)
+            nS = 4 * 12
+                    # Cliff Location
+            cliff = np.zeros(shape, dtype=bool)
+            cliff[3, 1:-1] = True
+
+            
+            hikers = [
```

Comment:
Should this to line 275 not be done in `init_render_state`

Reply:
fixed